teuthology/task/install/__init__.py

```diff
@@ -215,6 +216,43 @@ def install(ctx, config):
                 'python-ceph']
         rpms = ['ceph-fuse', 'librbd1', 'librados2', 'ceph-test', 'python-ceph']
     package_list = dict(deb=debs, rpm=rpms)
+
+    source_branch = config.get('branch')  # n-1 | n-2
+    if source_branch == 'n-1' or 'n-2':
+        import json
+        import subprocess
+        curlurl = 'curl -s  https://chacra.ceph.com/repos/ceph-release/ | jq "keys"'
+        status, output = subprocess.getstatusoutput(curlurl)
```

Comment:
This should not (or does not need to) be a dynamic part of the teuthology install task. We can construct the list of releases to upgrade from when creating the job matrix. Then it can be an explicit part of the job config / install task.

I recommend we table this for now until I can spend some time on the teuthology code to dynamically add dimensions to the matrix.

Reply:
> We can construct the list of releases to upgrade from when creating the job matrix

then would incur manual efforts which completely kills the purpose, the chacra API maintains the list, why not use it instead of constructing an additional one?

Reply:
It is not manual effort. We're moving the dynamic release lookup to the matrix generation. The advantage there is we can construct a matrix that upgrades for multiple minor releases of `n-1` or `n-2`, not just the current `HEAD`.

Reply:
Ah I remember from the talk post-standup a few weeks back, so the idea is to move even one step forward than this and not just to limit to few major releases but rigorously carry out tests on multiple minor releases, makes sense but the code you highlighted to be not part of the teuthology code:
> Then it can be an explicit part of the job config / install task.

would mean to get this into the YAMLs which would complicate the process, if there is some sort of abstraction i.e. to only have some placeholders in places in configs/install tasks like `n_1_minor_releases` would mean to fetch all the reef releases(just for instance keeping squid as ref point) and create the dynamic matrix you're referring to. What do you think?

Reply:
> Ah I remember from the talk post-standup a few weeks back, so the idea is to move even one step forward than this and not just to limit to few major releases but rigorously carry out tests on multiple minor releases, makes sense but the code you highlighted to be not part of the teuthology code:
> 
> > Then it can be an explicit part of the job config / install task.
> 
> would mean to get this into the YAMLs which would complicate the process, if there is some sort of abstraction i.e. to only have some placeholders in places in configs/install tasks like `n_1_minor_releases` would mean to fetch all the reef releases(just for instance keeping squid as ref point) and create the dynamic matrix you're referring to. What do you think?

You don't need placeholders. Just use an override for the `install:` task config.

Reply:
an override would have to simple like the ones we have right now like for instance:
```
overrides:
  ceph:
    conf:
```

doesn't that means we'd still have to get the code to get the dynamic thing done in the teuthology code itself? Otherwise how we code it explicitly? Lua in YAMLs?

Reply:
> doesn't that means we'd still have to get the code to get the dynamic thing done in the teuthology code itself?

teuthology-suite would adjust a directory like

`upgrade/tasks/0-from/{install, {base/{squid,reef}}}`

where `install.yaml` is something vanilla like

`install: {}`

and `reef.yaml` would have specially interpreted Lua code that expand to `v18.2.0, v18.2.1, v18.2.2, reef`.

Once it reaches the premerge/postmerge step, these dynamically created fragments already exist (and that's why we can't use the existing premerge/postmerge framework).